### PR TITLE
fix(agw): Migrate imsi_mme_ue_id_htbl in mme_app to protobuf map

### PR DESCRIPTION
**Pull request by pruthvihebbani**
fix(agw): Migrate imsi_mme_ue_id_htbl in mme_app to protobuf map

## Summary

Migrated imsi_mme_ue_id_htbl in mme_ap to protobuf map

## Test Plan

s1ap integ_tests
mme unit test
bazel build --config=production //lte/gateway/c/core:agw_of
bazel test //lte/gateway/c/core/oai/test/amf:amf_stateless_test
bazel test //lte/gateway/c/core/...
export FEATURES=mme_oai; make build_oai

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 1]
```diff
@@ -344,16 +344,15 @@ ue_mm_context_t* mme_ue_context_exists_mme_ue_s1ap_id(
 struct ue_mm_context_s* mme_ue_context_exists_imsi(
     mme_ue_context_t* const mme_ue_context_p, imsi64_t imsi) {
   hashtable_rc_t h_rc = HASH_TABLE_OK;
```
Comment (ajahl) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp]: this variable is not used

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 2]
```diff
@@ -575,14 +565,22 @@ void mme_ue_context_update_coll_keys(
   OAILOG_FUNC_OUT(LOG_MME_APP);
 }
 
+static bool display_proto_map_uint64_uint32(uint64_t keyP, const uint32_t dataP,
+                                            void* argP, void** resultP) {
+  OAILOG_DEBUG(LOG_MME_APP, "imsi2mme_ueid_map key=%llu, data=%lu\n", keyP,
+               dataP);
```
Comment (ajahl) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp]:
```diff
-OAILOG_DEBUG(LOG_MME_APP, "imsi2mme_ueid_map key=%llu, data=%lu\n", keyP,
-               dataP);
+OAILOG_DEBUG(LOG_MME_APP, "imsi2mme_ueid_map key=%lu, data=%u\n", keyP,
+               dataP);
```

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp [hunk 3]
```diff
@@ -700,22 +700,23 @@ status_code_e emm_context_upsert_imsi(emm_data_t* emm_data,
       (PARENT_STRUCT(elm, struct ue_mm_context_s, emm_context))->mme_ue_s1ap_id;
```
Comment (ajahl) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp:676]: `h_rc` is not used here
https://github.com/magma/magma/blob/8d67c44a8ab7085551932d5db47b77c89a6f400d/lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp#L698

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_manager.cpp [hunk 4]
```diff
@@ -30,7 +30,7 @@ constexpr char MME_NAS_STATE_KEY[] = "mme_nas_state";
 const int NUM_MAX_UE_HTBL_LISTS = 6;
 constexpr char UE_ID_UE_CTXT_TABLE_NAME[] =
     "mme_app_mme_ue_s1ap_id_ue_context_htbl";
-constexpr char IMSI_UE_ID_TABLE_NAME[] = "mme_app_imsi_ue_context_htbl";
+constexpr char MME_IMSI2MME_UE_ID[] = "mme_imsi2ue_id_map";
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_manager.cpp]: Change the name to "MME_IMSI2MME_UE_ID_MAP_NAME" to indicate it's map and it's name

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp [hunk 5]
```diff
@@ -700,22 +700,23 @@ status_code_e emm_context_upsert_imsi(emm_data_t* emm_data,
...
-      mme_app_desc_p->mme_ue_contexts.imsi_mme_ue_id_htbl,
-      (const hash_key_t)elm->_imsi64);
+  mme_app_desc_p->mme_ue_contexts.imsi2mme_ueid_map.remove(elm->_imsi64);
   if (INVALID_MME_UE_S1AP_ID != ue_id) {
-    h_rc = hashtable_uint64_ts_insert(
-        mme_app_desc_p->mme_ue_contexts.imsi_mme_ue_id_htbl,
-        (const hash_key_t)elm->_imsi64, ue_id);
+    if (mme_app_desc_p->mme_ue_contexts.imsi2mme_ueid_map.insert(
+            elm->_imsi64, ue_id) != magma::PROTO_MAP_OK) {
+      OAILOG_WARNING(LOG_MME_APP,
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp]: Use OAILOG_WARNING_UE

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp [hunk 6]
```diff
@@ -700,22 +700,23 @@ status_code_e emm_context_upsert_imsi(emm_data_t* emm_data,
...
   } else {
-    h_rc = HASH_TABLE_KEY_NOT_EXISTS;
-  }
-  if (HASH_TABLE_OK != h_rc) {
-    OAILOG_TRACE(LOG_MME_APP,
-                 "Error could not update this ue context "
-                 "mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT " imsi " IMSI_64_FMT
-                 ": %s\n",
-                 ue_id, elm->_imsi64, hashtable_rc_code2string(h_rc));
+    OAILOG_WARNING(LOG_MME_APP,
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp]: Use OAILOG_WARNING_UE

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 7]
```diff
@@ -344,16 +344,15 @@ ue_mm_context_t* mme_ue_context_exists_mme_ue_s1ap_id(
...
-    return mme_ue_context_exists_mme_ue_s1ap_id(
-        (mme_ue_s1ap_id_t)mme_ue_s1ap_id64);
+  if (INVALID_MME_UE_S1AP_ID != mme_ue_s1ap_id) {
+    return mme_ue_context_exists_mme_ue_s1ap_id(mme_ue_s1ap_id);
   } else {
-    OAILOG_WARNING_UE(LOG_MME_APP, imsi, " No IMSI hashtable for this IMSI\n");
+    OAILOG_WARNING_UE(LOG_MME_APP, imsi,
+                      " mme_ue_s1ap_id not found for this IMSI\n");
   }
   return NULL;
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp]: Change to nullptr

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 8]
```diff
@@ -575,14 +565,22 @@ void mme_ue_context_update_coll_keys(
   OAILOG_FUNC_OUT(LOG_MME_APP);
 }
 
+static bool display_proto_map_uint64_uint32(uint64_t keyP, const uint32_t dataP,
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp]: For unused arguments, set as 
__attribute__((unused)) void* argP

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 9]
```diff
@@ -676,12 +674,9 @@ status_code_e mme_insert_ue_context(
...
-          mme_ue_context_p->imsi_mme_ue_id_htbl,
-          (const hash_key_t)ue_context_p->emm_context._imsi64,
-          ue_context_p->mme_ue_s1ap_id);
-
-      if (HASH_TABLE_OK != h_rc) {
+      if (mme_ue_context_p->imsi2mme_ueid_map.insert(
+              ue_context_p->emm_context._imsi64,
+              ue_context_p->mme_ue_s1ap_id) != magma::PROTO_MAP_OK) {
         OAILOG_WARNING_UE(LOG_MME_APP, ue_context_p->emm_context._imsi64,
                           "Error could not register this ue context %p "
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp]: Can we log, Failed to insert imsi64 key to imsi2mme_ueid_map

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp [hunk 10]
```diff
@@ -706,9 +706,8 @@ void MmeNasStateConverter::state_to_proto(const mme_app_desc_t* mme_nas_state_p,
   auto mme_ue_ctxts_proto = state_proto->mutable_mme_ue_contexts();
 
   OAILOG_DEBUG(LOG_MME_APP, "IMSI table to proto");
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp]: IMSI map to proto map

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp [hunk 11]
```diff
@@ -746,8 +745,8 @@ void MmeNasStateConverter::proto_to_state(const oai::MmeNasState& state_proto,
   mme_ue_context_t* mme_ue_ctxt_state = &mme_nas_state_p->mme_ue_contexts;
   // copy maps to hashtables
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp]: Change the comment and log message

Comment (pruthvihebbani) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp]: There is a log at line 748

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 12]
```diff
@@ -677,12 +677,12 @@ status_code_e mme_insert_ue_context(
       if (mme_ue_context_p->imsi2mme_ueid_map.insert(
               ue_context_p->emm_context._imsi64,
               ue_context_p->mme_ue_s1ap_id) != magma::PROTO_MAP_OK) {
-        OAILOG_WARNING_UE(LOG_MME_APP, ue_context_p->emm_context._imsi64,
-                          "Error could not register this ue context %p "
-                          "mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT
-                          " imsi " IMSI_64_FMT "\n",
-                          ue_context_p, ue_context_p->mme_ue_s1ap_id,
-                          ue_context_p->emm_context._imsi64);
+        OAILOG_WARNING_UE(
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp:678]: imsi64 is already logged. So no need at the end

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp (hunk 11 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp]: Yes the log is : OAILOG_INFO(LOG_MME_APP, "Hashtable MME UE ID => IMSI"); 
We need to remove hashtable and put map 
The comment could be "copy in-memory map to protobuf map"

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp [hunk 13]
```diff
@@ -704,19 +703,19 @@ status_code_e emm_context_upsert_imsi(emm_data_t* emm_data,
   if (INVALID_MME_UE_S1AP_ID != ue_id) {
     if (mme_app_desc_p->mme_ue_contexts.imsi2mme_ueid_map.insert(
             elm->_imsi64, ue_id) != magma::PROTO_MAP_OK) {
-      OAILOG_WARNING(LOG_MME_APP,
-                     "Insert failed on imsi2mme_ueid_map "
-                     "mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT " imsi " IMSI_64_FMT
-                     "\n",
-                     ue_id, elm->_imsi64);
+      OAILOG_WARNING_UE(LOG_MME_APP, elm->_imsi64,
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp:682]: Imsi64 is already logged not required at the end

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp [hunk 14]
```diff
@@ -704,19 +703,19 @@ status_code_e emm_context_upsert_imsi(emm_data_t* emm_data,
...
     }
   } else {
-    OAILOG_WARNING(LOG_MME_APP,
-                   "Could not update ue context due to invalid "
-                   "mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT " imsi " IMSI_64_FMT
-                   "\n",
-                   ue_id, elm->_imsi64);
+    OAILOG_WARNING_UE(LOG_MME_APP, elm->_imsi64,
+                      "Could not update ue context due to invalid "
+                      "mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT " imsi " IMSI_64_FMT
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp]: \n is not required

**Review by rsarwad — APPROVED** [review]
LGTM

**Review by ajahl — APPROVED** [review]
LGTM

**Pull request merged**